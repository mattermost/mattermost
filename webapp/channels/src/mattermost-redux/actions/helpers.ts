// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import {ActionTypes} from 'mattermost-redux/action_types';
import {GenericAction} from 'mattermost-redux/types/actions';

export function getSharedChannels(): GenericAction {
    return {
        type: ActionTypes.GET_SHARED_CHANNELS,
    };
}

export function receivedSharedChannelsWithRemotes(sharedChannelsWithRemotes: any[]): GenericAction {
    return {
        type: ActionTypes.RECEIVED_SHARED_CHANNELS_WITH_REMOTES,
        data: sharedChannelsWithRemotes,
    };
}