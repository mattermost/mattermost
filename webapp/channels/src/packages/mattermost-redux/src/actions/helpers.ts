// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import type {AnyAction} from 'redux';

import type {ServerError} from '@mattermost/types/errors';
import type {GlobalState} from '@mattermost/types/store';

import {UserTypes} from 'mattermost-redux/action_types';
import {Client4} from 'mattermost-redux/client';
import type {DispatchFunc, GetStateFunc, ActionFuncAsync} from 'mattermost-redux/types/actions';

import {logError} from './errors';

type ActionType = string;
const HTTP_UNAUTHORIZED = 401;
export function forceLogoutIfNecessary(err: ServerError, dispatch: DispatchFunc<AnyAction>, getState: GetStateFunc) {
    const {currentUserId} = getState().entities.users;

    if ('status_code' in err && err.status_code === HTTP_UNAUTHORIZED && err.url && err.url.indexOf('/login') === -1 && currentUserId) {
        Client4.setToken('');
        dispatch({type: UserTypes.LOGOUT_SUCCESS, data: {}});
    }
}

function dispatcher(type: ActionType, data: any, dispatch: DispatchFunc<AnyAction>) {
    if (type.indexOf('SUCCESS') === -1) { // we don't want to pass the data for the request types
        dispatch(requestSuccess(type, data));
    } else {
        dispatch(requestData(type));
    }
}

export function requestData(type: ActionType) {
    return {
        type,
        data: null,
    };
}

export function requestSuccess(type: ActionType, data: any) {
    return {
        type,
        data,
    };
}

export function requestFailure(type: ActionType, error: ServerError): any {
    return {
        type,
        error,
    };
}

/**
 * Returns an ActionFunc which calls a specfied (client) function and
 * dispatches the specifed actions on request, success or failure.
 *
 * @export
 * @param {Object} obj                                       an object for destructirung required properties
 * @param {() => Promise<mixed>} obj.clientFunc              clientFunc to execute
 * @param {ActionType} obj.onRequest                         ActionType to dispatch on request
 * @param {(ActionType | Array<ActionType>)} obj.onSuccess   ActionType to dispatch on success
 * @param {ActionType} obj.onFailure                         ActionType to dispatch on failure
 * @param {...Array<any>} obj.params
 * @returns {ActionFunc} ActionFunc
 */

export function bindClientFunc<Func extends(...args: any[]) => Promise<any>>({
    clientFunc,
    onRequest,
    onSuccess,
    onFailure,
    params,
}: {
    clientFunc: Func;
    onRequest?: ActionType;
    onSuccess?: ActionType | ActionType[];
    onFailure?: ActionType;
    params?: Parameters<Func>;
}): ActionFuncAsync<Awaited<ReturnType<Func>>, GlobalState, AnyAction> {
    return async (dispatch, getState) => {
        if (onRequest) {
            dispatch(requestData(onRequest));
        }

        let data: Awaited<ReturnType<Func>>;
        try {
            if (params) {
                data = await clientFunc(...params);
            } else {
                data = await clientFunc();
            }
        } catch (error) {
            forceLogoutIfNecessary(error, dispatch, getState);
            if (onFailure) {
                dispatch(requestFailure(onFailure, error));
            }
            dispatch(logError(error));
            return {error};
        }

        if (Array.isArray(onSuccess)) {
            onSuccess.forEach((s) => {
                dispatcher(s, data, dispatch);
            });
        } else if (onSuccess) {
            dispatcher(onSuccess, data, dispatch);
        }

        return {data};
    };
}

// Debounce function based on underscores modified to use es6 and a cb

export function debounce(func: (...args: any) => unknown, wait: number, immediate?: boolean, cb?: () => unknown) {
    let timeout: NodeJS.Timeout|null;
    return function fx(...args: any[]) {
        const runLater = () => {
            timeout = null;
            if (!immediate) {
                Reflect.apply(func, null, args);
                if (cb) {
                    cb();
                }
            }
        };
        const callNow = immediate && !timeout;
        if (timeout) {
            clearTimeout(timeout);
        }
        timeout = setTimeout(runLater, wait);
        if (callNow) {
            Reflect.apply(func, null, args);
            if (cb) {
                cb();
            }
        }
    };
}

export class FormattedError extends Error {
    intl: {
        id: string;
        defaultMessage: string;
        values: any;
    };

    constructor(id: string, defaultMessage: string, values: any = {}) {
        super(defaultMessage);
        this.intl = {
            id,
            defaultMessage,
            values,
        };
    }
}

