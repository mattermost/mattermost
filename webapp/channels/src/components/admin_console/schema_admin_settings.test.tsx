// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import React from 'react';
import {FormattedMessage} from 'react-intl';

import type {CloudState} from '@mattermost/types/cloud';
import type {AdminConfig, EnvironmentConfig} from '@mattermost/types/config';

import SchemaText from 'components/admin_console/schema_text';

import {shallowWithIntl} from 'tests/helpers/intl-test-helper';

import SchemaAdminSettings from './schema_admin_settings';
import type {SchemaAdminSettings as SchemaAdminSettingsClass} from './schema_admin_settings';
import type {ConsoleAccess, AdminDefinitionSubSectionSchema, AdminDefinitionSettingInput} from './types';
import ValidationResult from './validation';

const DefaultProps = {
    cloud: {} as CloudState,
    consoleAccess: {} as ConsoleAccess,
    editRole: jest.fn(),
    enterpriseReady: false,
    isCurrentUserSystemAdmin: false,
    isDisabled: false,
    license: {},
    roles: {},
    setNavigationBlocked: jest.fn(),
};

describe('components/admin_console/SchemaAdminSettings', () => {
    let schema: AdminDefinitionSubSectionSchema | null = null;
    let config: Partial<AdminConfig> = {};
    let environmentConfig: Partial<EnvironmentConfig> = {};

    afterEach(() => {
        schema = null;
        config = {};
        environmentConfig = {};
    });

    beforeEach(() => {
        schema = {
            id: 'Config',
            name: 'config',
            name_default: 'Configuration',
            settings: [
                {
                    key: 'FirstSettings.settinga',
                    label: 'label-a',
                    label_default: 'Setting One',
                    type: 'text',
                    default: 'setting_default',
                    help_text: 'help-text-a',
                    help_text_default: 'This is some help text for the text field.',
                    placeholder: 'placeholder-a',
                    placeholder_default: 'e.g. some setting',
                },
                {
                    key: 'FirstSettings.settingb',
                    label: 'label-b',
                    label_default: 'Setting Two',
                    type: 'bool',
                    default: true,
                    help_text: 'help-text-b',
                    help_text_default: 'This is some help text for the bool field.',
                },
                {
                    key: 'FirstSettings.settingc',
                    label: 'label-c',
                    label_default: 'Setting Three',
                    type: 'dropdown',
                    default: 'option1',
                    options: [
                        {display_name: 'Option 1', value: 'option1'},
                        {display_name: 'Option 2', value: 'option2'},
                        {display_name: 'Option 3', value: 'option3'},
                    ],
                    help_text: 'help-text-c',
                    help_text_default: 'This is some help text for the dropdown field.',
                },
                {
                    key: 'SecondSettings.settingd',
                    label: 'label-d',
                    label_default: 'Setting Four',
                    type: 'radio',
                    default: 'option2',
                    options: [
                        {display_name: 'Option 1', value: 'option1'},
                        {display_name: 'Option 2', value: 'option2'},
                        {display_name: 'Option 3', value: 'option3'},
                    ],
                    help_text: 'help-text-d',
                    help_text_default: 'This is some help text for the radio field.',
                },
                {
                    key: 'SecondSettings.settinge',
                    label: 'label-e',
                    label_default: 'Setting Five',
                    type: 'generated',
                    help_text: 'help-text-e',
                    help_text_default: 'This is some help text for the generated field.',
                    regenerate_help_text: 'This is help text for the regenerate button.',
                    placeholder: 'placeholder-e',
                    placeholder_default: 'e.g. 47KyfOxtk5+ovi1MDHFyzMDHIA6esMWb',
                },
                {
                    key: 'SecondSettings.settingf',
                    label: 'label-f',
                    label_default: 'Setting Six',
                    type: 'username',
                    help_text: 'help-text-f',
                    help_text_default: 'This is some help text for the user autocomplete field.',
                    placeholder: 'placeholder-f',
                    placeholder_default: 'Type a username here',
                },
                {
                    key: 'SecondSettings.settingg',
                    label: 'label-g',
                    label_default: 'Setting Seven',
                    type: 'number',
                    default: 'setting_default',
                    help_text: 'help-text-g',
                    help_text_default: 'This is some help text for the number field.',
                    placeholder: 'placeholder-g',
                    placeholder_default: 'e.g. some setting',
                },
                {
                    key: 'SecondSettings.settingh',
                    label: 'label-h',
                    label_default: 'Setting Eight',
                    type: 'number',
                    default: 'setting_default',
                    help_text: 'help-text-h',
                    help_text_default: 'This is some help text for the number field.',
                    placeholder: 'placeholder-h',
                    placeholder_default: 'e.g. some setting',
                    onConfigLoad: (configVal) => configVal / 10,
                    onConfigSave: (displayVal) => displayVal * 10,
                },
                {
                    label: 'label-h',
                    label_default: 'Setting Eight',
                    type: 'banner',
                },
                {
                    key: 'SecondSettings.settingi',
                    label: 'label-i',
                    label_default: 'Setting Nine',
                    type: 'language',
                    help_text: 'help-text-i',
                    help_text_default: 'This is some help text for the language field.',
                    placeholder: 'placeholder-i',
                    placeholder_default: 'e.g. some setting',
                    multiple: false,
                },
                {
                    key: 'SecondSettings.settingj',
                    label: 'label-j',
                    label_default: 'Setting Nine',
                    type: 'language',
                    help_text: 'help-text-j',
                    help_text_default: 'This is some help text for the multiple-language field.',
                    placeholder: 'placeholder-j',
                    placeholder_default: 'e.g. some setting',
                    multiple: true,
                    no_result: 'no-result-j',
                    no_result_default: 'No result',
                    not_present: 'no-present-j',
                    not_present_default: 'No present',
                },
                {
                    key: 'SecondSettings.settingk',
                    label: 'label-k',
                    label_default: 'Setting Eleven',
                    type: 'button',
                    help_text: 'help-text-k',
                    help_text_default: 'This is some help text for the button field.',
                    action: () => null,
                    error_message: 'admin.reload.reloadFail',
                    error_message_default: 'Reload unsuccessful: {error}',
                },
                {
                    key: 'FirstSettings.settingl',
                    label: 'label-l',
                    label_default: 'Setting Twelve',
                    type: 'bool',
                    default: false,
                    help_text: 'help-text-l',
                    help_text_default: 'This is some help text for the second bool field.',
                },
                {
                    key: 'FirstSettings.settingm',
                    label: 'label-m',
                    label_default: 'Setting Thirteen',
                    type: 'color',
                    help_text: 'help-text-m',
                    help_text_default: 'This is some help text for the color field.',
                },
                {
                    type: 'custom',
                    key: 'custom',
                    component: () => <p>{'Test'}</p>,
                },
                {
                    type: 'jobstable',
                    label: 'label-l',
                    label_default: 'Setting Twelve',
                    help_text: 'help-text-l',
                    help_text_default: 'This is some help text for the jobs table field.',
                    job_type: 'test',
                    render_job: () => <p>{'Test'}</p>,
                },
                {
                    key: 'EscapedSettings.com+example+setting.a',
                    label: 'escaped-label-a',
                    label_default: 'Escaped Setting A',
                    type: 'bool',
                    default: false,
                    help_text: 'escaped-help-text-a',
                    help_text_default: 'This is some help text for the first escaped field.',
                },
            ],
        } as AdminDefinitionSubSectionSchema;
        config = {
            FirstSettings: {
                settinga: 'fsdsdg',
                settingb: false,
                settingc: 'option3',
                settingl: true,
            },
            SecondSettings: {
                settingd: 'option1',
                settinge: 'Q6DHXrFLOIS5sOI5JNF4PyDLqWm7vh23',
                settingf: '3xz3r6n7dtbbmgref3yw4zg7sr',
                settingg: 7,
                settingh: 100,
            },
            EscapedSettings: {
                'com.example.setting': {
                    a: true,
                },
            },
        } as Partial<AdminConfig>;
        environmentConfig = {
            FirstSettings: {
                settingl: true,
            },
        } as Partial<EnvironmentConfig>;
    });

    test('should match snapshot with settings and plugin', () => {
        const wrapper = shallowWithIntl(
            <SchemaAdminSettings
                {...DefaultProps}
                config={config}
                environmentConfig={environmentConfig}
                schema={{...schema} as AdminDefinitionSubSectionSchema}
                patchConfig={jest.fn()}
            />,
        );
        expect(wrapper).toMatchSnapshot();
    });

    test('should match snapshot with custom component', () => {
        const wrapper = shallowWithIntl(
            <SchemaAdminSettings
                {...DefaultProps}
                config={config}
                environmentConfig={environmentConfig}
                schema={{component: () => <p>{'Test'}</p>} as AdminDefinitionSubSectionSchema}
                patchConfig={jest.fn()}
            />,
        );
        expect(wrapper).toMatchSnapshot();
    });

    test('should render header using a SchemaText', () => {
        const headerText = 'This is [a link](!https://example.com) in the header';
        const props = {
            ...DefaultProps,
            config,
            environmentConfig,
            schema: {
                ...schema,
                header: headerText,
            } as AdminDefinitionSubSectionSchema,
            patchConfig: jest.fn(),
        };

        const wrapper = shallowWithIntl(<SchemaAdminSettings {...props}/>);

        const header = wrapper.find(SchemaText);
        expect(header.exists()).toBe(true);
        expect(header.props()).toMatchObject({
            text: headerText,
            isMarkdown: true,
        });
    });

    test('should render footer using a SchemaText', () => {
        const footerText = 'This is [a link](https://example.com) in the footer';
        const props = {
            ...DefaultProps,
            config,
            environmentConfig,
            schema: {
                ...schema,
                footer: footerText,
            } as AdminDefinitionSubSectionSchema,
            patchConfig: jest.fn(),
        };

        const wrapper = shallowWithIntl(<SchemaAdminSettings {...props}/>);

        const footer = wrapper.find(SchemaText);
        expect(footer.exists()).toBe(true);
        expect(footer.props()).toMatchObject({
            text: footerText,
            isMarkdown: true,
        });
    });

    test('should render page not found', () => {
        const props = {
            ...DefaultProps,
            config,
            environmentConfig,
            schema: null,
            patchConfig: jest.fn(),
        };

        const wrapper = shallowWithIntl(<SchemaAdminSettings {...props}/>);

        expect(wrapper.contains(
            <FormattedMessage
                id='error.plugin_not_found.title'
                defaultMessage='Plugin Not Found'
            />,
        )).toEqual(true);
    });

    test('should not try to validate when a setting does not contain a key', () => {
        const mockValidate = jest.fn(() => {
            return new ValidationResult(true, '');
        });

        const localSchema = {...schema} as AdminDefinitionSubSectionSchema & {settings: AdminDefinitionSettingInput[]};
        localSchema.settings = [
            {
                label: 'a banner', // won't validate because no key
                type: 'banner' as any,
                validate: mockValidate,
            },
        ];

        const props = {
            ...DefaultProps,
            config,
            id: '',
            environmentConfig,
            schema: localSchema,
            patchConfig: jest.fn(),
        };

        const wrapper = shallowWithIntl(<SchemaAdminSettings {...props}/>);
        const instance = wrapper.instance() as SchemaAdminSettingsClass;

        expect(instance.canSave()).toBe(true);
        expect(mockValidate).not.toHaveBeenCalled();
    });

    test('should validate when a setting contains a key and a validation method', () => {
        const mockValidate = jest.fn(() => {
            return new ValidationResult(true, '');
        });

        const localSchema = {...schema} as AdminDefinitionSubSectionSchema & {settings: AdminDefinitionSettingInput[]};
        localSchema.settings = [
            {
                key: 'field1', // will validate because it has a key AND a validate method
                label: 'with key and validation',
                type: 'text',
                validate: mockValidate,
            },
        ];
        const props = {
            ...DefaultProps,
            config,
            environmentConfig,
            schema: localSchema,
            patchConfig: jest.fn(),
        };

        const wrapper = shallowWithIntl(<SchemaAdminSettings {...props}/>);
        const instance = wrapper.instance() as SchemaAdminSettingsClass;

        expect(instance.canSave()).toBe(true);
        expect(mockValidate).toHaveBeenCalled();
    });
});
