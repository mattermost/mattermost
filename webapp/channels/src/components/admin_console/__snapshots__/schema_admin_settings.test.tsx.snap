// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/admin_console/SchemaAdminSettings should match snapshot with custom component 1`] = `
<component
  cloud={Object {}}
  config={
    Object {
      "EscapedSettings": Object {
        "com.example.setting": Object {
          "a": true,
        },
      },
      "FirstSettings": Object {
        "settinga": "fsdsdg",
        "settingb": false,
        "settingc": "option3",
        "settingl": true,
      },
      "SecondSettings": Object {
        "settingd": "option1",
        "settinge": "Q6DHXrFLOIS5sOI5JNF4PyDLqWm7vh23",
        "settingf": "3xz3r6n7dtbbmgref3yw4zg7sr",
        "settingg": 7,
        "settingh": 100,
      },
    }
  }
  consoleAccess={Object {}}
  disabled={false}
  editRole={[MockFunction]}
  enterpriseReady={false}
  environmentConfig={
    Object {
      "FirstSettings": Object {
        "settingl": true,
      },
    }
  }
  intl={
    Object {
      "$t": [Function],
      "defaultFormats": Object {},
      "defaultLocale": "en",
      "defaultRichTextElements": undefined,
      "fallbackOnEmptyString": true,
      "formatDate": [Function],
      "formatDateTimeRange": [Function],
      "formatDateToParts": [Function],
      "formatDisplayName": [Function],
      "formatList": [Function],
      "formatListToParts": [Function],
      "formatMessage": [Function],
      "formatNumber": [Function],
      "formatNumberToParts": [Function],
      "formatPlural": [Function],
      "formatRelativeTime": [Function],
      "formatTime": [Function],
      "formatTimeToParts": [Function],
      "formats": Object {},
      "formatters": Object {
        "getDateTimeFormat": [Function],
        "getDisplayNames": [Function],
        "getListFormat": [Function],
        "getMessageFormat": [Function],
        "getNumberFormat": [Function],
        "getPluralRules": [Function],
        "getRelativeTimeFormat": [Function],
      },
      "locale": "en",
      "messages": Object {},
      "onError": [Function],
      "onWarn": [Function],
      "textComponent": "span",
      "timeZone": "Etc/UTC",
    }
  }
  isCurrentUserSystemAdmin={false}
  isDisabled={false}
  license={Object {}}
  patchConfig={[MockFunction]}
  roles={Object {}}
  schema={
    Object {
      "component": [Function],
    }
  }
  setNavigationBlocked={[MockFunction]}
/>
`;

exports[`components/admin_console/SchemaAdminSettings should match snapshot with settings and plugin 1`] = `
<div
  className="wrapper--fixed "
>
  <AdminHeader>
    config
  </AdminHeader>
  <div
    className="admin-console__wrapper"
  >
    <div
      className="admin-console__content"
    >
      <form
        className="form-horizontal"
        onSubmit={[Function]}
        role="form"
      >
        <Memo(SettingsGroup)
          container={false}
        >
          <AdminTextSetting
            disabled={false}
            footer={null}
            helpText={
              <SchemaText
                text="help-text-a"
              />
            }
            id="FirstSettings.settinga"
            key="Config_text_FirstSettings.settinga"
            label="label-a"
            onChange={[Function]}
            placeholder="placeholder-a"
            setByEnv={false}
            type="text"
            value="fsdsdg"
          />
          <Memo(BooleanSetting)
            disabled={false}
            helpText={
              <SchemaText
                text="help-text-b"
              />
            }
            id="FirstSettings.settingb"
            key="Config_bool_FirstSettings.settingb"
            label="label-b"
            onChange={[Function]}
            setByEnv={false}
            value={false}
          />
          <Memo(DropdownSetting)
            disabled={false}
            helpText={
              <SchemaText
                text="help-text-c"
              />
            }
            id="FirstSettings.settingc"
            key="Config_dropdown_FirstSettings.settingc"
            label="label-c"
            onChange={[Function]}
            setByEnv={false}
            value="option3"
            values={
              Array [
                Object {
                  "text": "Option 1",
                  "value": "option1",
                },
                Object {
                  "text": "Option 2",
                  "value": "option2",
                },
                Object {
                  "text": "Option 3",
                  "value": "option3",
                },
              ]
            }
          />
          <Memo(RadioSetting)
            disabled={false}
            helpText={
              <SchemaText
                text="help-text-d"
              />
            }
            id="SecondSettings.settingd"
            key="Config_radio_SecondSettings.settingd"
            label="label-d"
            onChange={[Function]}
            setByEnv={false}
            value="option1"
            values={
              Array [
                Object {
                  "text": "Option 1",
                  "value": "option1",
                },
                Object {
                  "text": "Option 2",
                  "value": "option2",
                },
                Object {
                  "text": "Option 3",
                  "value": "option3",
                },
              ]
            }
          />
          <GeneratedSetting
            disabled={false}
            helpText={
              <SchemaText
                text="help-text-e"
              />
            }
            id="SecondSettings.settinge"
            key="Config_generated_SecondSettings.settinge"
            label="label-e"
            onChange={[Function]}
            placeholder="placeholder-e"
            regenerateHelpText="This is help text for the regenerate button."
            regenerateText={
              <Memo(MemoizedFormattedMessage)
                defaultMessage="Regenerate"
                id="admin.regenerate"
              />
            }
            setByEnv={false}
            value="Q6DHXrFLOIS5sOI5JNF4PyDLqWm7vh23"
          />
          <Connect(UserAutocompleteSetting)
            disabled={false}
            helpText={
              <SchemaText
                text="help-text-f"
              />
            }
            id="SecondSettings.settingf"
            key="Config_userautocomplete_SecondSettings.settingf"
            label="label-f"
            onChange={[Function]}
            placeholder="placeholder-f"
            value="3xz3r6n7dtbbmgref3yw4zg7sr"
          />
          <AdminTextSetting
            disabled={false}
            footer={null}
            helpText={
              <SchemaText
                text="help-text-g"
              />
            }
            id="SecondSettings.settingg"
            key="Config_text_SecondSettings.settingg"
            label="label-g"
            onChange={[Function]}
            placeholder="placeholder-g"
            setByEnv={false}
            type="number"
            value={7}
          />
          <AdminTextSetting
            disabled={false}
            footer={null}
            helpText={
              <SchemaText
                text="help-text-h"
              />
            }
            id="SecondSettings.settingh"
            key="Config_text_SecondSettings.settingh"
            label="label-h"
            onChange={[Function]}
            placeholder="placeholder-h"
            setByEnv={false}
            type="number"
            value={10}
          />
          <div
            className="banner undefined"
            key="Config_bool_undefined"
          >
            <div
              className="banner__content"
            >
              <span>
                <span>
                  label-h
                </span>
              </span>
            </div>
          </div>
          <Memo(DropdownSetting)
            disabled={false}
            helpText={
              <SchemaText
                text="help-text-i"
              />
            }
            id="SecondSettings.settingi"
            key="Config_language_SecondSettings.settingi"
            label="label-i"
            onChange={[Function]}
            setByEnv={false}
            value="de"
            values={
              Array [
                Object {
                  "order": 0,
                  "text": "Deutsch",
                  "value": "de",
                },
                Object {
                  "order": 1,
                  "text": "English (US)",
                  "value": "en",
                },
                Object {
                  "order": 2,
                  "text": "English (Australia)",
                  "value": "en-AU",
                },
                Object {
                  "order": 3,
                  "text": "Español (Alpha)",
                  "value": "es",
                },
                Object {
                  "order": 4,
                  "text": "Français (Beta)",
                  "value": "fr",
                },
                Object {
                  "order": 5,
                  "text": "Italiano (Alpha)",
                  "value": "it",
                },
                Object {
                  "order": 6,
                  "text": "Magyar (Beta)",
                  "value": "hu",
                },
                Object {
                  "order": 7,
                  "text": "Nederlands",
                  "value": "nl",
                },
                Object {
                  "order": 8,
                  "text": "Polski",
                  "value": "pl",
                },
                Object {
                  "order": 9,
                  "text": "Português (Brasil) (Alpha)",
                  "value": "pt-BR",
                },
                Object {
                  "order": 10,
                  "text": "Română (Alpha)",
                  "value": "ro",
                },
                Object {
                  "order": 11,
                  "text": "Svenska",
                  "value": "sv",
                },
                Object {
                  "order": 12,
                  "text": "Tiếng Việt (Beta)",
                  "value": "vi",
                },
                Object {
                  "order": 13,
                  "text": "Türkçe",
                  "value": "tr",
                },
                Object {
                  "order": 14,
                  "text": "Български (Alpha)",
                  "value": "bg",
                },
                Object {
                  "order": 15,
                  "text": "Pусский",
                  "value": "ru",
                },
                Object {
                  "order": 16,
                  "text": "Yкраїнська (Alpha)",
                  "value": "uk",
                },
                Object {
                  "order": 17,
                  "text": "فارسی (Alpha)",
                  "value": "fa",
                },
                Object {
                  "order": 18,
                  "text": "한국어 (Alpha)",
                  "value": "ko",
                },
                Object {
                  "order": 19,
                  "text": "中文 (简体) (Beta)",
                  "value": "zh-CN",
                },
                Object {
                  "order": 20,
                  "text": "中文 (繁體) (Beta)",
                  "value": "zh-TW",
                },
                Object {
                  "order": 21,
                  "text": "日本語",
                  "value": "ja",
                },
              ]
            }
          />
          <Memo(MultiSelectSetting)
            disabled={false}
            helpText={
              <SchemaText
                text="help-text-j"
              />
            }
            id="SecondSettings.settingj"
            key="Config_language_SecondSettings.settingj"
            label="label-j"
            noResultText="no-result-j"
            onChange={[Function]}
            selected={Array []}
            setByEnv={false}
            values={
              Array [
                Object {
                  "order": 0,
                  "text": "Deutsch",
                  "value": "de",
                },
                Object {
                  "order": 1,
                  "text": "English (US)",
                  "value": "en",
                },
                Object {
                  "order": 2,
                  "text": "English (Australia)",
                  "value": "en-AU",
                },
                Object {
                  "order": 3,
                  "text": "Español (Alpha)",
                  "value": "es",
                },
                Object {
                  "order": 4,
                  "text": "Français (Beta)",
                  "value": "fr",
                },
                Object {
                  "order": 5,
                  "text": "Italiano (Alpha)",
                  "value": "it",
                },
                Object {
                  "order": 6,
                  "text": "Magyar (Beta)",
                  "value": "hu",
                },
                Object {
                  "order": 7,
                  "text": "Nederlands",
                  "value": "nl",
                },
                Object {
                  "order": 8,
                  "text": "Polski",
                  "value": "pl",
                },
                Object {
                  "order": 9,
                  "text": "Português (Brasil) (Alpha)",
                  "value": "pt-BR",
                },
                Object {
                  "order": 10,
                  "text": "Română (Alpha)",
                  "value": "ro",
                },
                Object {
                  "order": 11,
                  "text": "Svenska",
                  "value": "sv",
                },
                Object {
                  "order": 12,
                  "text": "Tiếng Việt (Beta)",
                  "value": "vi",
                },
                Object {
                  "order": 13,
                  "text": "Türkçe",
                  "value": "tr",
                },
                Object {
                  "order": 14,
                  "text": "Български (Alpha)",
                  "value": "bg",
                },
                Object {
                  "order": 15,
                  "text": "Pусский",
                  "value": "ru",
                },
                Object {
                  "order": 16,
                  "text": "Yкраїнська (Alpha)",
                  "value": "uk",
                },
                Object {
                  "order": 17,
                  "text": "فارسی (Alpha)",
                  "value": "fa",
                },
                Object {
                  "order": 18,
                  "text": "한국어 (Alpha)",
                  "value": "ko",
                },
                Object {
                  "order": 19,
                  "text": "中文 (简体) (Beta)",
                  "value": "zh-CN",
                },
                Object {
                  "order": 20,
                  "text": "中文 (繁體) (Beta)",
                  "value": "zh-TW",
                },
                Object {
                  "order": 21,
                  "text": "日本語",
                  "value": "ja",
                },
              ]
            }
          />
          <RequestButton
            buttonText={
              <span>
                label-k
              </span>
            }
            disabled={false}
            errorMessage="admin.reload.reloadFail"
            helpText={
              <SchemaText
                text="help-text-k"
              />
            }
            id="SecondSettings.settingk"
            includeDetailedError={true}
            key="Config_text_SecondSettings.settingk"
            requestAction={[Function]}
            saveNeeded={false}
            showSuccessMessage={false}
            successMessage={
              Object {
                "defaultMessage": "Test Successful",
                "id": "admin.requestButton.requestSuccess",
              }
            }
          />
          <Memo(BooleanSetting)
            disabled={false}
            helpText={
              <SchemaText
                text="help-text-l"
              />
            }
            id="FirstSettings.settingl"
            key="Config_bool_FirstSettings.settingl"
            label="label-l"
            onChange={[Function]}
            setByEnv={true}
            value={true}
          />
          <Memo(ColorSetting)
            disabled={false}
            helpText={
              <SchemaText
                text="help-text-m"
              />
            }
            id="FirstSettings.settingm"
            key="Config_text_FirstSettings.settingm"
            label="label-m"
            onChange={[Function]}
            value=""
          />
          <component
            cancelSubmit={[Function]}
            config={
              Object {
                "EscapedSettings": Object {
                  "com.example.setting": Object {
                    "a": true,
                  },
                },
                "FirstSettings": Object {
                  "settinga": "fsdsdg",
                  "settingb": false,
                  "settingc": "option3",
                  "settingl": true,
                },
                "SecondSettings": Object {
                  "settingd": "option1",
                  "settinge": "Q6DHXrFLOIS5sOI5JNF4PyDLqWm7vh23",
                  "settingf": "3xz3r6n7dtbbmgref3yw4zg7sr",
                  "settingg": 7,
                  "settingh": 100,
                },
              }
            }
            disabled={false}
            helpText={
              <span>
                
              </span>
            }
            id="custom"
            key="Config_custom_custom"
            label=""
            license={Object {}}
            onChange={[Function]}
            registerSaveAction={[Function]}
            setByEnv={false}
            setSaveNeeded={[Function]}
            showConfirm={false}
            unRegisterSaveAction={[Function]}
          />
          <Connect(JobTable)
            createJobButtonText="label-l"
            createJobHelpText={
              <SchemaText
                text="help-text-l"
              />
            }
            disabled={false}
            getExtraInfoText={[Function]}
            jobType="test"
            key="Config_jobstable_undefined"
          />
          <Memo(BooleanSetting)
            disabled={false}
            helpText={
              <SchemaText
                text="escaped-help-text-a"
              />
            }
            id="EscapedSettings.com+example+setting.a"
            key="Config_bool_EscapedSettings.com+example+setting.a"
            label="escaped-label-a"
            onChange={[Function]}
            setByEnv={false}
            value={true}
          />
        </Memo(SettingsGroup)>
      </form>
    </div>
  </div>
  <div
    className="admin-console-save"
  >
    <SaveButton
      disabled={true}
      onClick={[Function]}
      saving={false}
      savingMessage="Saving Config..."
    />
    <WithTooltip
      id="error-tooltip"
      placement="top"
      title=""
    >
      <div
        className="error-message"
        data-testid="errorMessage"
      >
        <Memo(FormError)
          error=""
          iconClassName="fa-exclamation-triangle"
          textClassName="has-warning"
        />
        <Memo(FormError)
          error={null}
        />
      </div>
    </WithTooltip>
  </div>
</div>
`;
